### ISSUE TITLE
sonar-config -i  languages requests.exceptions.HTTPError: 400 Client Error

### ISSUE THREAD

**Issue by Jay-jakwak4** [okorach/sonar-tools#1447]
Hi Team
There is an issue with sonar-config -i  :  languages 

it's our current value. 
```
"globalSettings" - "languages"

"sonar.cfamily.generateComputedConfig": false,


"sonar.cfamily.ignoreHeaderComments": true,


"sonar.java.springconfig.file.patterns": "**/src/main/resources/**/application*.properties, **/src/main/resources/**/application*.yaml, **/src/main/resources/**/application*.yml"
```

```
2024-10-24 16:24:36,686 | sonar-config | ERROR   | MainThread      | URL https://sonarqube-secondary.****.com/api/settings/set?key=sonar.cfamily.generateComputedConfig&value=false: HTTP error 400 - : Setting 'sonar.cfamily.generateComputedConfig' cannot be global ( request)
--- Logging error ---
Traceback (most recent call last):
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/settings.py", line 505, in set_setting
    s.set(value)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/settings.py", line 256, in set
    return self.post(API_SET, params=params).ok
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/sqobject.py", line 102, in post
    return self.endpoint.post(api=api, params=params, exit_on_error=exit_on_error, mute=mute)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 201, in post
    return self.__run_request(requests.post, api, params, exit_on_error, mute, **kwargs)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 256, in __run_request
    raise e
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 249, in __run_request
    r.raise_for_status()
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/requests/models.py", line 1024, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error:  for url: https://sonarqube-secondary.****.com/api/settings/set?key=sonar.cfamily.generateComputedConfig&value=false

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/logging/__init__.py", line 1083, in emit
    msg = self.format(record)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/logging/__init__.py", line 927, in format
    return fmt.format(record)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/logging/__init__.py", line 663, in format
    record.message = record.getMessage()
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/logging/__init__.py", line 367, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "/Users/jakwak4/Library/Python/3.9/bin/sonar-config", line 8, in <module>
    sys.exit(main())
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/cli/config.py", line 342, in main
    __import_config(endpoint, what, **kwargs)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/cli/config.py", line 310, in __import_config
    func(endpoint, data, key_list=key_list)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 853, in import_config
    endpoint.import_config(config_data)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 476, in import_config
    self.set_setting(s, v)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 373, in set_setting
    return settings.set_setting(self, key, value)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/settings.py", line 507, in set_setting
    log.error("%s while setting setting '%s'", util.error_msg(e), key, str(component))
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/logging.py", line 96, in error
    __LOGGER.error(*params)
Message: "%s while setting setting '%s'"
Arguments: ("URL https://sonarqube-secondary.****.com/api/settings/set?key=sonar.cfamily.generateComputedConfig&value=false: HTTP error 400 - : Setting 'sonar.cfamily.generateComputedConfig' cannot be global", 'sonar.cfamily.generateComputedConfig', 'None')
2024-10-24 16:24:36,811 | sonar-config | ERROR   | MainThread      | URL https://sonarqube-secondary.****.com/api/settings/set?key=sonar.cfamily.ignoreHeaderComments&value=true: HTTP error 400 - : Setting 'sonar.cfamily.ignoreHeaderComments' cannot be global ( request)
--- Logging error ---
Traceback (most recent call last):
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/settings.py", line 505, in set_setting
    s.set(value)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/settings.py", line 256, in set
    return self.post(API_SET, params=params).ok
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/sqobject.py", line 102, in post
    return self.endpoint.post(api=api, params=params, exit_on_error=exit_on_error, mute=mute)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 201, in post
    return self.__run_request(requests.post, api, params, exit_on_error, mute, **kwargs)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 256, in __run_request
    raise e
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 249, in __run_request
    r.raise_for_status()
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/requests/models.py", line 1024, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error:  for url: https://sonarqube-secondary.****.com/api/settings/set?key=sonar.cfamily.ignoreHeaderComments&value=true

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/logging/__init__.py", line 1083, in emit
    msg = self.format(record)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/logging/__init__.py", line 927, in format
    return fmt.format(record)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/logging/__init__.py", line 663, in format
    record.message = record.getMessage()
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/logging/__init__.py", line 367, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "/Users/jakwak4/Library/Python/3.9/bin/sonar-config", line 8, in <module>
    sys.exit(main())
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/cli/config.py", line 342, in main
    __import_config(endpoint, what, **kwargs)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/cli/config.py", line 310, in __import_config
    func(endpoint, data, key_list=key_list)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 853, in import_config
    endpoint.import_config(config_data)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 476, in import_config
    self.set_setting(s, v)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 373, in set_setting
    return settings.set_setting(self, key, value)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/settings.py", line 507, in set_setting
    log.error("%s while setting setting '%s'", util.error_msg(e), key, str(component))
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/logging.py", line 96, in error
    __LOGGER.error(*params)
Message: "%s while setting setting '%s'"
Arguments: ("URL https://sonarqube-secondary.****.com/api/settings/set?key=sonar.cfamily.ignoreHeaderComments&value=true: HTTP error 400 - : Setting 'sonar.cfamily.ignoreHeaderComments' cannot be global", 'sonar.cfamily.ignoreHeaderComments', 'None')
2024-10-24 16:24:39,595 | sonar-config | ERROR   | MainThread      | URL https://sonarqube-secondary.****.com/api/settings/set?key=sonar.java.springconfig.file.patterns&value=%2A%2A%2Fsrc%2Fmain%2Fresources%2F%2A%2A%2Fapplication%2A.properties%2C+%2A%2A%2Fsrc%2Fmain%2Fresources%2F%2A%2A%2Fapplication%2A.yaml%2C+%2A%2A%2Fsrc%2Fmain%2Fresources%2F%2A%2A%2Fapplication%2A.yml: HTTP error 400 - : Parameter 'value' must be used for single value setting. Parameter 'values' must be used for multi value setting. ( request)
--- Logging error ---
Traceback (most recent call last):
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/settings.py", line 505, in set_setting
    s.set(value)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/settings.py", line 256, in set
    return self.post(API_SET, params=params).ok
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/sqobject.py", line 102, in post
    return self.endpoint.post(api=api, params=params, exit_on_error=exit_on_error, mute=mute)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 201, in post
    return self.__run_request(requests.post, api, params, exit_on_error, mute, **kwargs)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 256, in __run_request
    raise e
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 249, in __run_request
    r.raise_for_status()
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/requests/models.py", line 1024, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error:  for url: https://sonarqube-secondary.****.com/api/settings/set?key=sonar.java.springconfig.file.patterns&value=%2A%2A%2Fsrc%2Fmain%2Fresources%2F%2A%2A%2Fapplication%2A.properties%2C+%2A%2A%2Fsrc%2Fmain%2Fresources%2F%2A%2A%2Fapplication%2A.yaml%2C+%2A%2A%2Fsrc%2Fmain%2Fresources%2F%2A%2A%2Fapplication%2A.yml

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/logging/__init__.py", line 1083, in emit
    msg = self.format(record)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/logging/__init__.py", line 927, in format
    return fmt.format(record)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/logging/__init__.py", line 663, in format
    record.message = record.getMessage()
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/logging/__init__.py", line 367, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "/Users/jakwak4/Library/Python/3.9/bin/sonar-config", line 8, in <module>
    sys.exit(main())
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/cli/config.py", line 342, in main
    __import_config(endpoint, what, **kwargs)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/cli/config.py", line 310, in __import_config
    func(endpoint, data, key_list=key_list)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 853, in import_config
    endpoint.import_config(config_data)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 476, in import_config
    self.set_setting(s, v)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/platform.py", line 373, in set_setting
    return settings.set_setting(self, key, value)
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/settings.py", line 507, in set_setting
    log.error("%s while setting setting '%s'", util.error_msg(e), key, str(component))
  File "/Users/jakwak4/Library/Python/3.9/lib/python/site-packages/sonar/logging.py", line 96, in error
    __LOGGER.error(*params)
Message: "%s while setting setting '%s'"
Arguments: ("URL https://sonarqube-secondary.****.com/api/settings/set?key=sonar.java.springconfig.file.patterns&value=%2A%2A%2Fsrc%2Fmain%2Fresources%2F%2A%2A%2Fapplication%2A.properties%2C+%2A%2A%2Fsrc%2Fmain%2Fresources%2F%2A%2A%2Fapplication%2A.yaml%2C+%2A%2A%2Fsrc%2Fmain%2Fresources%2F%2A%2A%2Fapplication%2A.yml: HTTP error 400 - : Parameter 'value' must be used for single value setting. Parameter 'values' must be used for multi value setting.", 'sonar.java.springconfig.file.patterns', 'None')
```

Thanks

**Comment by okorach** [comment]
Hello @Jay-jakwak4 ,

Thanks for reporting this:
About the error on `sonar.java.springconfig.file.patterns` this is the same problem as #1442 and will be fixed globally
About the error on `sonar.cfamily.ignoreHeaderComments` and `sonar.cfamily.generateComputedConfig` did you add yourself the settings in the global settings section or it is the result of an `sonar-config` export?
I am asking because this setting can't be set (and therefore should not be exported) in global settings. I can only be set per project.
Did you not patch the global settings JSON to add this property?

I will make sure to harden that part, ie if you set a global setting that can't be set, I will catch the error, but this is something that should not happen from a well formed JSON file

**Comment by okorach** [comment]
OK, I think that I found the root cause. It's just the log on the error 400 (that I also have myself) that causes the stack trace.
I will fix that. BTW, I will remove from global settings, parameters from the export that can be GET but not SET globally (weird behavior of SonarQube on this one)

**Comment by Jay-jakwak4** [comment]
Thank you for your answer. 
I will try to test again when the next version is released

**Comment by okorach** [comment]
Fixed in #1455
